### DISCUSSION TITLE
aws-elasticloadbalancingv2: How to achieve path-based and weighted target group routing from ALB to multiple services?

### DISCUSSION THREAD

**Discussion by gokulanv** [aws/aws-cdk#28334] [Q&A]
### Describe the issue

I have service A with two pipeline prod stages: 1-box serves 10% traffic and full-prod that serves 90% traffic.
Now, I want to add Service B to the same cluster and implement path based routing.

```
ALB 
        -> Path: /svcA -> TargetGroup-SvcA-1Box: Service A (1-box stage 10% traffic)
        -> Path: /svcA -> TargetGroup-SvcA-FullProd: Service A (full-prod stage 90% traffic)
        -> Path: /svcB -> TargetGroup-SvcB-1Box: Service B (1-box stage 10% traffic)
        -> Path: /svcB -> TargetGroup-SvcB-FullProd: Service B (full-prod stage 90% traffic)
```
Can you point me to CDK examples of how I can do this? I dont see enough documentation to support both path-based and weight based service routing from ALB. At the bare minimum, would like to know step by step guidelines with constructs to use.

### Links

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/listener-update-rules.html

**Comment by tim-finnigan** [comment]
Hi @gokulanv thanks for reaching out. Here is the CDK documentation for `aws-elasticloadbalancingv2`: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2-readme.html. 

You can find more info there for your use case, for example:

```ts
declare const listener: elbv2.ApplicationListener;
declare const asg: autoscaling.AutoScalingGroup;

listener.addTargets('Example.Com Fleet', {
  priority: 10,
  conditions: [
    elbv2.ListenerCondition.hostHeaders(['example.com']),
    elbv2.ListenerCondition.pathPatterns(['/ok', '/path']),
  ],
  port: 8080,
  targets: [asg]
});
```
Also:
>* **Weighted Target Groups:** use ListenerAction.weightedForward() to give different weights to different target groups.

I hope that helps. I'm going to convert this issue into a discussion since it involves a general question. Others from the community are welcome to share feedback here as well.

**Reply by gokulanv** [reply]
I did it this way: Is this correct?
```
const albListener = this.applicationLoadBalancer.addListener('Listener-1', {
      protocol: ApplicationProtocol.HTTPS,
      port: props.listenerPortFleet,
      certificates: [{ certificateArn: this.certificate.certificateArn }],
      defaultAction: ListenerAction.fixedResponse(200, {
        contentType: 'text/plain',
        messageBody: 'OK',
      }),
    });

const applicationListenerRuleSps = new ApplicationListenerRule(this, 'SvcA-Rule', {
      listener: albListener,
      priority: 1,
      conditions: [ListenerCondition.pathPatterns(['/api'])],
      action: ListenerAction.weightedForward(loadBalancerTargetGroupsforServiceA),
    });
    const applicationListenerRule = new ApplicationListenerRule(this, 'SvcB-Rule', {
      listener: albListener,
      priority: 2,
      conditions: [ListenerCondition.pathPatterns(['/events'])],
      action: ListenerAction.weightedForward(loadBalancerTargetGroupsforServiceB),
    });
```
